From 2d2ab1450995c87065254e8492270dba51a18892 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Fri, 10 Sep 2010 00:44:54 -0700
Subject: [PATCH] Oprofile: Octeon: Change IRQ flags to IRQ_NODELAY when threaded IRQ enabled

The original oprofile IRQ flags is IRQF_SHARED. And, on Octeon, orpofile
IRQ is a common external interrupt. So, CONFIG_PREEMPT_HARDIRQS will
thread it. This cause oprofile IRQ doesn't work well with threaded IRQ.
So, change the IRQ flags to IRQ_NODELAY.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 arch/mips/oprofile/op_model_cavium_octeon.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/arch/mips/oprofile/op_model_cavium_octeon.c b/arch/mips/oprofile/op_model_cavium_octeon.c
index a910443..969ddd0 100644
--- a/arch/mips/oprofile/op_model_cavium_octeon.c
+++ b/arch/mips/oprofile/op_model_cavium_octeon.c
@@ -114,8 +114,14 @@ static irqreturn_t octeon_perfcount_handler(int irq, void *dev_id)
 
 static int octeon_init(void)
 {
+	unsigned long irq_flags;
+#ifdef CONFIG_PREEMPT_HARDIRQS
+	irq_flags = IRQF_NODELAY;
+#else
+	irq_flags = IRQF_SHARED;
+#endif 
 	int result =
-		request_irq(OCTEON_IRQ_PERF, octeon_perfcount_handler, IRQF_SHARED,
+		request_irq(OCTEON_IRQ_PERF, octeon_perfcount_handler, irq_flags,
 			    "Perfcounter", octeon_perfcount_handler);
 #ifdef CONFIG_SMP
 	if (result == 0) {
-- 
1.6.5.2

